compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -851,20 +861,41 @@ impl<'hir> LoweringContext<'_, 'hir> {
         whole_span: Span,
     ) -> hir::ExprKind<'hir> {
         // Return early in case of an ordinary assignment.
-        fn is_ordinary(lhs: &Expr) -> bool {
+        fn is_ordinary(lower_ctx: &mut LoweringContext<'_, '_>, lhs: &Expr) -> bool {
             match &lhs.kind {
-                ExprKind::Tup(..) => false,
+                ExprKind::Array(..) | ExprKind::Struct(..) | ExprKind::Tup(..) => false,
+                ExprKind::Call(callee, ..) => {
+                    // Check for tuple struct constructor.
+                    if let ExprKind::Path(qself, path) = &callee.kind {
+                        let qpath = lower_ctx.lower_qpath(
+                            callee.id,
+                            qself,
+                            path,
+                            ParamMode::Optional,
+                            ImplTraitContext::disallowed(),
+                        );
+                        match qpath {
+                            hir::QPath::Resolved(
+                                _,
+                                hir::Path { res: Res::Def(DefKind::Ctor(..), _), .. },
```

Comment:
I suggest to keep the rules syntactic and do not look at resolutions, like in all other cases of destructuring.

The `DefKind::Ctor(..)` check is both too relaxed (the correct condition is `DefKind::Ctor(_, CtorKind::Fn)`), and, what is worse, too restrictive, because this check will reject cases like these
```rust
Self(a, b) = ...;

enum E { V(u8, u16) }
type A = E;
A::V(a, b) = ...;
```
for which we don't know the resolution during AST lowering, and determine it only during type checking.

Reply:
It would be good to add test cases for the examples above.
And also a test case for a free function call, non-path call, and qualified path call in this position.
```
fn foo(a: u8, b: u16) {}

foo(a, b) = ...;
(x)(a, b) = ...;
<Type as Trait>::foo(a, b) = ...;
```

Reply:
I changed it as you suggested. I feel that this made some error messages worse (see test cases). Let me know what you think.

Reply:
I think we can keep the resolution-based check in a slightly different form for diagnostics.

If `self.resolver.get_partial_res(id)` returns `Some(partial_res)` and `partial_res.unresolved_segments() == 0` and `!matches!(partial_res.base_res(), Res::Def(DefKind::Ctor(_, CtorKind::Fn), _) | Res::SelfCtor(..))`

https://github.com/rust-lang/rust/blob/87a0997ef9c0bfad0ba362741afa601d8fb285b8/compiler/rustc_resolve/src/late.rs#L301-L303

then we are sure that it's not a tuple struct and can treat it as a regular assignment, otherwise defer to type checking.

This will need a comment telling that it's for diagnostics and a link to `PathSource::is_expected` telling to keep the logic in sync with that function.

Reply:
That sounds like a good solution. I was just wondering whether instead of a comment to keep things in sync, we could just call `PathSource::TupleStruct(lhs.span, &args.map(|e| e.span).collect::<Vec<_>>()).is_expected()` or something like that directly to avoid duplication.

Reply:
`rustc_ast_lowering` and `rustc_resolve` do not depend on each other but they both depend on `rustc_hir`, so you could share the `matches!(res, Res::Def(DefKind::Ctor(_, CtorKind::Fn), _) | Res::SelfCtor(..))` part by moving it to a method on `Res`, `expected_in_tuple_struct_pat` or something.

Reply:
Thanks! I've tried your idea with `get_partial_res(id)` and pushed it. However, 1 test fails: for `Alias::SingleVariant(...) = ...` in `src/test/ui/destructuring-assignment/tuple_struct_destructure.rs`, I get `Some(PartialRes { base_res: Def(TyAlias, DefId(0:10)), unresolved_segments: 1 })` as the partial resolution. Hence this case is not recognized correctly. I don't understand why the resolver cannot figure this out.